### PR TITLE
Use FlutterProject to locate files 

### PR DESCRIPTION
**Pull request by sigurdm**
A bit of cleanup of flutter_tools' handling of project directories and manifests.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/create.dart [hunk 1]
```diff
@@ -362,16 +361,16 @@ To edit platform code in an IDE see https://flutter.io/developing-packages/#edit
...
+    return template.render(directory, context, overwriteExisting: false);
   }
 
-  int _injectGradleWrapper(String projectDir) {
+  int _injectGradleWrapper(FlutterProject project) {
     int filesCreated = 0;
     copyDirectorySync(
       cache.getArtifactDirectory('gradle_wrapper'),
-      fs.directory(fs.path.join(projectDir, 'android')),
+      project.directory.childDirectory('android'),
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/commands/create.dart]: `project.android`?

File: packages/flutter_tools/lib/src/ios/cocoapods.dart [hunk 2]
```diff
@@ -173,17 +173,16 @@ class CocoaPods {
...
+    _addPodsDependencyToFlutterXcconfig(iosProject, 'Release');
   }
 
-  void _addPodsDependencyToFlutterXcconfig(String appDirectory, String mode) {
-    final File file = fs.file(fs.path.join(appDirectory, 'ios', 'Flutter', '$mode.xcconfig'));
+  void _addPodsDependencyToFlutterXcconfig(IosProject iosProject, String mode) {
+    final File file = iosProject.xcodeConfigFor(mode);
     if (file.existsSync()) {
-      final String content = file.readAsStringSync();
+      final String content = iosProject.xcodeConfigFor(mode).readAsStringSync();
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/ios/cocoapods.dart]: Could we not retain the use of `file` here?

File: packages/flutter_tools/lib/src/ios/cocoapods.dart [hunk 3]
```diff
@@ -206,24 +204,25 @@ class CocoaPods {
...
     if (dependenciesChanged)
       return true;
-    final File podfileFile = appIosDirectory.childFile('Podfile');
-    final File podfileLockFile = appIosDirectory.childFile('Podfile.lock');
-    final File manifestLockFile =
-        appIosDirectory.childFile(fs.path.join('Pods', 'Manifest.lock'));
+
+    final File podfileFile = iosProject.podfile;
+    final File podfileLockFile = iosProject.podfileLock;
+    final File manifestLockFile =iosProject.podManifestLock;
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/ios/cocoapods.dart]: Add space after `=`

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart [hunk 4]
```diff
@@ -25,27 +28,18 @@ String flutterFrameworkDir(BuildMode mode) {
...
 /// Writes default Xcode properties files in the Flutter project at [projectPath],
 /// if project is an iOS project and such files are out of date or do not
 /// already exist.
-void generateXcodeProperties({String projectPath, FlutterManifest manifest}) {
-  if (manifest.isModule || fs.isDirectorySync(fs.path.join(projectPath, 'ios'))) {
-    final File propertiesFile = fs.file(_generatedXcodePropertiesPath(projectPath: projectPath, manifest: manifest));
-    if (!Cache.instance.fileOlderThanToolsStamp(propertiesFile)) {
+Future<void> generateXcodeProperties({FlutterProject project}) async {
+  if ((await project.manifest).isModule ||
+      project.directory.childDirectory('ios').existsSync()) {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/ios/xcodeproj.dart]: `project.ios.directory.existsSync()`?

File: packages/flutter_tools/lib/src/plugins.dart [hunk 5]
```diff
@@ -155,8 +158,18 @@ void _writeAndroidPluginRegistrant(String directory, List<Plugin> plugins) {
...
 
-  final String javaSourcePath = fs.path.join(directory, 'src', 'main', 'java');
-  final String registryPath = fs.path.join(javaSourcePath, 'io', 'flutter', 'plugins', 'GeneratedPluginRegistrant.java');
+  final String javaSourcePath = fs.path.join(
+    (await project.androidPluginRegistrantDirectory).path,
+    'src',
+    'main',
+    'java',
+  );
+  final String registryPath = fs.path.join(javaSourcePath,
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/plugins.dart]: [nit] Move `javaSourcePath,` to the next line.

File: packages/flutter_tools/lib/src/plugins.dart [hunk 6]
```diff
@@ -255,57 +266,37 @@ void _writeIOSPluginRegistrant(String directory, FlutterManifest manifest, List<
...
-  if (manifest.isModule) {
-    _writeAndroidPluginRegistrant(fs.path.join(projectPath, '.android', 'Flutter'), plugins);
-  } else if (fs.isDirectorySync(fs.path.join(projectPath, 'android', 'app'))) {
-    _writeAndroidPluginRegistrant(fs.path.join(projectPath, 'android', 'app'), plugins);
-  }
-  if (manifest.isModule) {
-    _writeIOSPluginRegistrant(fs.path.join(projectPath, '.ios'), manifest, plugins);
-  } else if (fs.isDirectorySync(fs.path.join(projectPath, 'ios'))) {
-    _writeIOSPluginRegistrant(fs.path.join(projectPath, 'ios'), manifest, plugins);
+Future<void> injectPlugins({@required FlutterProject project}) async {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/plugins.dart]: Maybe this method should just take a plain, positional parameter.

File: packages/flutter_tools/lib/src/project.dart [hunk 7]
```diff
@@ -67,6 +67,57 @@ class FlutterProject {
...
+  Future<File> get generatedXcodePropertiesFile {
+    return _generatedXcodeProperties ??= manifest.then((FlutterManifest manifest) {
+      return directory.childDirectory(manifest.isModule ? '.ios' : 'ios')
+          .childDirectory('Flutter')
+          .childFile('Generated.xcconfig');
+    });
+  }
+  Future<File> _generatedXcodeProperties;
+
+  File get flutterPluginsDotFile {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/project.dart]: This property name seems a bit strange to me, as the `.` is a file name prefix. Maybe remove `Dot`?

File: packages/flutter_tools/lib/src/project.dart [hunk 8]
```diff
@@ -67,6 +67,57 @@ class FlutterProject {
...
+  Future<File> get androidLocalPropertiesFile {
+    return _androidLocalPropertiesFile ??= manifest.then((FlutterManifest manifest) {
+      return directory.childDirectory(manifest.isModule ? '.android' : 'android')
+          .childFile('local.properties');
+    });
+  }
+  Future<File> _androidLocalPropertiesFile;
+
+  Future<File> get generatedXcodePropertiesFile {
+    return _generatedXcodeProperties ??= manifest.then((FlutterManifest manifest) {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/project.dart]: Add `<File>` after `then`

File: packages/flutter_tools/lib/src/project.dart [hunk 9]
```diff
@@ -67,6 +67,57 @@ class FlutterProject {
   /// The generated IosModule sub project of this module project.
   IosModuleProject get iosModule => new IosModuleProject(directory.childDirectory('.ios'));
 
+  Future<File> get androidLocalPropertiesFile {
+    return _androidLocalPropertiesFile ??= manifest.then((FlutterManifest manifest) {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/project.dart]: Add `<File>` after `then`.

File: packages/flutter_tools/lib/src/project.dart [hunk 10]
```diff
@@ -67,6 +67,57 @@ class FlutterProject {
...
+    });
+  }
+  Future<File> _generatedXcodeProperties;
+
+  File get flutterPluginsDotFile {
+    return _flutterPluginsDotFile ??= directory.childFile('.flutter-plugins');
+  }
+  File _flutterPluginsDotFile;
+
+  Future<Directory> get androidPluginRegistrantDirectory async {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/project.dart]: This `Directory` is not where we should place the registrant. It is the Android sub project in which we should place the registrant. This discrepancy might be confusing.

Maybe `androidPluginHost` or similar?

File: packages/flutter_tools/lib/src/project.dart [hunk 11]
```diff
@@ -67,6 +67,57 @@ class FlutterProject {
...
+      if (manifest.isModule) {
+        return directory.childDirectory('.android').childDirectory('Flutter');
+      } else {
+        return directory.childDirectory('android').childDirectory('app');
+      }
+    });
+  }
+  Future<Directory> _androidPluginRegistrantDirectory;
+
+  Future<Directory> get iosPluginRegistrantDirectory async {
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/project.dart]: Similarly here.

File: packages/flutter_tools/test/project_test.dart [hunk 12]
```diff
@@ -17,8 +17,7 @@ void main() {
       final Directory directory = fs.directory('myproject');
       expect(
         new FlutterProject(directory).directory.absolute.path,
-        directory.absolute.path,
-      );
+        directory.absolute.path,);
```
Comment (mravn-google) [inline packages/flutter_tools/test/project_test.dart]: Please put the `);` back to whence it came.

**Review by mravn-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/commands/create.dart (hunk 1 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/commands/create.dart]: Done

File: packages/flutter_tools/lib/src/ios/cocoapods.dart (hunk 2 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/ios/cocoapods.dart]: Yes!
Done

File: packages/flutter_tools/lib/src/ios/cocoapods.dart (hunk 3 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/ios/cocoapods.dart]: Done

File: packages/flutter_tools/lib/src/ios/xcodeproj.dart (hunk 4 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/ios/xcodeproj.dart]: Done

File: packages/flutter_tools/lib/src/plugins.dart (hunk 5 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/plugins.dart]: Done

File: packages/flutter_tools/lib/src/plugins.dart (hunk 6 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/plugins.dart]: Done

File: packages/flutter_tools/lib/src/project.dart (hunk 9 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/project.dart]: Done

File: packages/flutter_tools/lib/src/project.dart (hunk 8 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/project.dart]: Done

File: packages/flutter_tools/lib/src/project.dart (hunk 7 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/project.dart]: Done

File: packages/flutter_tools/lib/src/project.dart (hunk 10 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/project.dart]: Better!
Done

File: packages/flutter_tools/lib/src/project.dart (hunk 11 above)
Comment (sigurdm) [reply packages/flutter_tools/lib/src/project.dart]: Done

File: packages/flutter_tools/test/project_test.dart (hunk 12 above)
Comment (sigurdm) [reply packages/flutter_tools/test/project_test.dart]: ok, ok... will do );

**Review by sigurdm — COMMENTED** [review]
Thanks for the review!

**Comment by sigurdm** [comment]
@szakarias Could you take a look. Especially the latest commit?

**Review by szakarias — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/application_package.dart [hunk 13]
```diff
@@ -31,29 +32,29 @@ abstract class ApplicationPackage {
...
+  File get packagesFile => null;
 
   @override
   String toString() => displayName;
 }
 
 class AndroidApk extends ApplicationPackage {
   /// Path to the actual apk file.
-  final String apkPath;
+  final File apk;
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply packages/flutter_tools/lib/src/application_package.dart]: Done

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 14]
```diff
@@ -23,9 +23,7 @@ import '../project.dart';
 import 'android_sdk.dart';
 import 'android_studio.dart';
 
-const String gradleManifestPath = 'android/app/src/main/AndroidManifest.xml';
-const String gradleAppOutV1 = 'android/app/build/outputs/apk/app-debug.apk';
-const String gradleAppOutDirV1 = 'android/app/build/outputs/apk';
+
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply packages/flutter_tools/lib/src/android/gradle.dart]: Removed

File: packages/flutter_tools/lib/src/application_package.dart [hunk 15]
```diff
@@ -306,20 +307,20 @@ class PrebuiltIOSApp extends IOSApp {
 }
 
 Future<ApplicationPackage> getApplicationPackageForPlatform(TargetPlatform platform, {
-  String applicationBinary
+  File apk
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply packages/flutter_tools/lib/src/application_package.dart]: Done

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 16]
```diff
@@ -271,7 +271,7 @@ class FlutterDevice {
     final TargetPlatform targetPlatform = await device.targetPlatform;
     package = await getApplicationPackageForPlatform(
       targetPlatform,
-      applicationBinary: hotRunner.applicationBinary
+      apk: hotRunner.applicationBinary
```
[the remark this answers is not in the document]
Comment (sigurdm) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Done

**Pull request merged**